### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Yevhenii0536**
[DEMO LINK](https://yevhenii0536.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/App.scss [hunk 1]
```diff
@@ -6,34 +6,41 @@ html {
...
-  &--completed &__title {
-    color: #080;
+  &--completed {
+    background-color: #c7fa9d5b;
   }
 }
 
 .UserInfo {
   font-style: italic;
 }
```
Comment (valeriiayarzhemska) [inline src/App.scss:33]: u need to move these in different files

File: src/App.tsx [hunk 2]
```diff
@@ -1,61 +1,122 @@
...
+
+function getUser(userId: number): User | null {
+  const foundUser = usersFromServer.find((user) => user.id === userId);
+
+  return foundUser || null;
+}
+
+function getMaximalId(todoes: Todo[]) {
+  return Math.max(...todoes.map((todo) => todo.id)) + 1;
+}
```
Comment (valeriiayarzhemska) [inline src/App.tsx:20]: add what does function return

File: src/App.tsx [hunk 3]
```diff
@@ -1,61 +1,122 @@
...
+  const resetState = () => {
+    setTitle('');
+    setUserId(0);
+    setIsHaveTitle(false);
+    setIsHaveUserId(false);
+  };
+
+  const addNewTodo = (event: React.ChangeEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    const newTodo: Todo = {
```
Comment (valeriiayarzhemska) [inline src/App.tsx:43]: add a row here

File: src/App.tsx [hunk 4]
```diff
@@ -1,61 +1,122 @@
...
+      user: getUser(userId),
+    };
+
+    if (title.length === 0) {
+      setIsHaveTitle(true);
+    }
+
+    if (userId < 1) {
+      setIsHaveUserId(true);
+    }
```
Comment (valeriiayarzhemska) [inline src/App.tsx:57]: u can check with '!'

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,122 @@
...
+
+    if (title.length > 0 && userId > 0) {
+      setNewTodo([...todoes, newTodo]);
+
+      resetState();
+    }
```
Comment (valeriiayarzhemska) [inline src/App.tsx:63]: u can remove '> 0'

File: src/App.tsx [hunk 6]
```diff
@@ -1,61 +1,122 @@
...
+    setIsHaveUserId(false);
+  };
+
   return (
     <div className="App">
-      <h1>Add todo form</h1>
+      <h1 className="App__title">Add todo form</h1>
 
-      <form action="/api/users" method="POST">
+      <form action="/api/users" method="POST" onSubmit={addNewTodo}>
```
Comment (valeriiayarzhemska) [inline src/App.tsx:82]: add on a new rows

File: src/App.tsx [hunk 7]
```diff
@@ -1,61 +1,122 @@
...
-          <span className="error">Please enter a title</span>
+          <Form.Control
+            id="title"
+            placeholder="Enter a title"
+            type="text"
+            data-cy="titleInput"
+            value={title}
+            onChange={handleChangeInput}
+          />
+          {isHaveTitle && <span className="error">Please enter a title</span>}
```
Comment (valeriiayarzhemska) [inline src/App.tsx:92]: add space

File: src/App.tsx [hunk 8]
```diff
@@ -1,61 +1,122 @@
...
+          >
+            <option disabled value={0}>
+              Choose a user
+            </option>
+            {usersFromServer.map((user) => (
+              <option key={user.id} value={user.id}>
+                {user.name}
+              </option>
+            ))}
+          </Form.Select>
```
Comment (valeriiayarzhemska) [inline src/App.tsx:109]: add space

File: src/App.tsx [hunk 9]
```diff
@@ -1,61 +1,122 @@
...
 
-          <span className="error">Please choose a user</span>
+          {isHaveUserId && <span className="error">Please choose a user</span>}
```
Comment (valeriiayarzhemska) [inline src/App.tsx:111]: add space

File: src/App.tsx [hunk 10]
```diff
@@ -1,61 +1,122 @@
...
         </div>
 
-        <button type="submit" data-cy="submitButton">
+        <Button variant="outline-success" type="submit" data-cy="submitButton">
```
Comment (valeriiayarzhemska) [inline src/App.tsx:114]: add space

**Review by valeriiayarzhemska — CHANGES_REQUESTED** [review]
gj, but some issues to fix:)

**Review by valeriiayarzhemska — APPROVED** [review]

File: src/App.tsx [hunk 11]
```diff
@@ -1,61 +1,143 @@
...
+import { useState } from 'react';
+import { User } from './types/User';
+import { Todo } from './types/Todo';
+import { TodoList } from './components/TodoList';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUser(userId: number): User | null {
+  const foundUser = usersFromServer.find((user) => user.id === userId);
```
Comment (Andr1yK) [inline src/App.tsx:14]:
```suggestion
  const user = usersFromServer.find((user) => user.id === userId);
```

File: src/App.tsx [hunk 12]
```diff
@@ -1,61 +1,143 @@
...
+    if (!title.length) {
+      setIsHaveTitle(true);
+    }
+
+    if (!userId) {
+      setIsHaveUserId(true);
+    }
+
+    if (title.length && userId) {
+      setNewTodo([...todoes, newTodo]);
```
Comment (Andr1yK) [inline src/App.tsx:62]: In ReactJS, it’s recommended to use functional `setState` instead of the regular `setState`. The reason for this is that the functional form of `setState` can help avoid issues with stale state, and ensure that the state is updated correctly.

Here are some rules to consider when using functional `setState` in React:

1. **Always use functional `setState` when updating state based on previous state:** If you need to update the state based on its previous value, it’s recommended to use functional `setState` to ensure that the state is updated correctly. Here’s an example:

```jsx
// using functional setState
this.setState(prevState => ({ count: prevState.count + 1 }));
```

2. **Use regular `setState` for simple updates:** If you’re updating the state with a simple value, such as a boolean or a string, you can use regular `setState`. Here’s an example:

```jsx
// using regular setState
this.setState({ isModalOpen: true });
```

3. **Use functional `setState` for complex state updates:** If you’re updating the state with a complex value, such as an array or an object, it’s recommended to use functional `setState` to ensure that the state is updated correctly. Here’s an example:

```jsx
// using functional setState with object spread syntax
this.setState(prevState => ({ ...prevState, items: newItems }));
```

4. **Be careful with nested state updates:** If you have nested state updates, it’s important to use functional `setState` to ensure that the state is updated correctly. Here’s an example:

```jsx
// incorrect use of setState with nested state updates
this.setState({ items: [] });
this.setState(prevState => ({ items: [...prevState.items, newItem] }));

// correct use of functional setState with nested state updates
this.setState(prevState => ({ items: [...prevState.items, newItem] }));
```

In summary, using functional `setState` in ReactJS can help avoid issues with stale state and ensure that the state is updated correctly. Always use functional `setState` when updating state based on its previous value, and be careful with nested state updates.

File: src/App.tsx [hunk 13]
```diff
@@ -1,61 +1,143 @@
...
+          <Form.Control
+            id="title"
+            placeholder="Enter a title"
+            type="text"
+            data-cy="titleInput"
+            value={title}
+            onChange={handleChangeInput}
+          />
+
+          {isHaveTitle && (
```
Comment (Andr1yK) [inline src/App.tsx:99]: is it the correct name for a variable?

File: src/App.tsx [hunk 14]
```diff
@@ -1,61 +1,143 @@
...
+            <option disabled value={0}>
+              Choose a user
+            </option>
+
+            {usersFromServer.map((user) => (
+              <option key={user.id} value={user.id}>
+                {user.name}
+              </option>
+            ))}
+          </Form.Select>
```
Comment (Andr1yK) [inline src/App.tsx:122]:
```suggestion
          >
            <option disabled value={0}>
              Choose a user
            </option>

            {usersFromServer.map((user) => (
              <option key={user.id} value={user.id}>
                {user.name}
              </option>
            ))}
          </Form.Select>
```

File: src/App.tsx [hunk 15]
```diff
@@ -1,61 +1,143 @@
...
 
-          <span className="error">Please choose a user</span>
+          {isHaveUserId && (
```
Comment (Andr1yK) [inline src/App.tsx:124]: the same here

File: src/components/TodoList/TodoList.tsx [hunk 16]
```diff
@@ -1 +1,20 @@
...
+};
+
+export const TodoList: React.FC<Props> = ({ todos }) => {
+  return (
+    <section className="TodoList">
+      {todos.map((todo) => {
+        return (
+          <TodoInfo todo={todo} key={todo.id} />
+        );
+      })}
```
Comment (Andr1yK) [inline src/components/TodoList/TodoList.tsx:17]:
```suggestion
      {todos.map((todo) => (
          <TodoInfo todo={todo} key={todo.id} />
      ))}
```

**Review by Andr1yK — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 17]
```diff
@@ -1,61 +1,148 @@
...
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUserById(userId: number): User | null {
+  const user = usersFromServer.find((person) => person.id === userId);
+
+  return user || null;
+}
+
+function getMaxId(todos: Todo[]): number {
```
Comment (sergii-nosachenko) [inline src/App.tsx:19]:
```suggestion
function getNextTodoId(todos: Todo[]): number {
```

File: src/App.tsx [hunk 18]
```diff
@@ -1,61 +1,148 @@
...
+      user: getUserById(userId),
+    };
+
+    if (!title.length) {
+      setIsTitleValid(true);
+    }
+
+    if (!userId) {
+      setIsUserIdValid(true);
+    }
```
Comment (sergii-nosachenko) [inline src/App.tsx:59]:
```suggestion
    if (!title.length || !userId) {
      setIsTitleValid(!title.length);
      setIsUserIdValid(!userId);
      
      return;
    }
```

Combine logic, and return from the handler to prevent the next action

File: src/App.tsx [hunk 19]
```diff
@@ -1,61 +1,148 @@
...
+
+    setNewTodo((prevTodos) => {
+      if (title.length && userId) {
```
Comment (sergii-nosachenko) [inline src/App.tsx:62]: Useless if you check it before

File: src/App.tsx [hunk 20]
```diff
@@ -1,61 +1,148 @@
...
+        const updatedTodos = [...prevTodos, newTodo];
+
+        resetForm();
```
Comment (sergii-nosachenko) [inline src/App.tsx:65]: Call it after setter, not inside

File: src/App.tsx [hunk 21]
```diff
@@ -1,61 +1,148 @@
...
+
+            {usersFromServer.map((user) => (
+              <option key={user.id} value={user.id}>
+                {user.name}
+              </option>
+            ))}
+          </Form.Select>
 
-          <span className="error">Please choose a user</span>
+          {isUserIdValid && (
```
Comment (sergii-nosachenko) [inline src/App.tsx:129]: If not valid

File: src/components/TodoInfo/TodoInfo.tsx [hunk 22]
```diff
@@ -1 +1,33 @@
...
+
+  return (
+    <article
+      data-id={id}
+      className={classNames(
+        'TodoInfo',
+        { 'TodoInfo--completed': completed },
+      )}
+    >
+      <h2 className="TodoInfo__title">{title}</h2>
```
Comment (sergii-nosachenko) [inline src/components/TodoInfo/TodoInfo.tsx:26]: Split into rows

File: src/components/TodoList/TodoList.tsx [hunk 23]
```diff
@@ -1 +1,18 @@
...
+import './TodoList.scss';
+
+type Props = {
+  todos: Todo[]
+};
+
+export const TodoList: React.FC<Props> = ({ todos }) => (
+  <section className="TodoList">
+    {todos.map((todo) => {
+      return (
```
Comment (sergii-nosachenko) [inline src/components/TodoList/TodoList.tsx:13]: Useless `return`

File: src/App.tsx [hunk 24]
```diff
@@ -27,60 +27,55 @@ export const initialTodos: Todo[] = [...todosFromServer].map((todo) => ({
 
 export const App = () => {
   const [title, setTitle] = useState('');
-  const [userId, setUserId] = useState<number>(0);
+  const [userId, setUserId] = useState(0);
   const [todos, setNewTodo] = useState(initialTodos);
-  const [isTitleValid, setIsTitleValid] = useState(false);
-  const [isUserIdValid, setIsUserIdValid] = useState(false);
+  const [isHasTitle, setIsHasTitle] = useState(false);
+  const [isUserIdInvalid, setIsUserIdInvalid] = useState(false);
```
Comment (sergii-nosachenko) [inline src/App.tsx:33]:
```suggestion
  const [hasTitle, setHasTitle] = useState(false);
  const [isUserIdInvalid, setIsUserIdInvalid] = useState(false);
```

**Review by sergii-nosachenko — APPROVED** [review]
Commented
